ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,476 @@
+    final boolean afterDelete;
+
+    OperationType(boolean nonBlockingRouter, boolean afterDelete) {
+      this.nonBlockingRouter = nonBlockingRouter;
+      this.afterDelete = afterDelete;
+    }
+  }
+
+  /**
+   * Describes an operation chain and provides useful metadata and operation results for testing.
+   */
+  static class OperationInfo {
+    final int operationId;
+    final BlobProperties properties;
+    final byte[] userMetadata;
+    final byte[] data;
+    final Queue<OperationType> opChain;
+    final List<TestFuture> futures = new ArrayList<>();
+    final CountDownLatch latch = new CountDownLatch(1);
+    String blobId;
+
+    OperationInfo(int operationId, BlobProperties properties, byte[] userMetadata, byte[] data,
+        Queue<OperationType> opChain) {
+      this.operationId = operationId;
+      this.properties = properties;
+      this.userMetadata = userMetadata;
+      this.data = data;
+      this.opChain = opChain;
+    }
+  }
+
+  /**
+   * This class encapsulates a future and allows the user to define a check method that runs tests on
+   * the retrieved value in the future.
+   *
+   * @param <T> The type of the encapsulated future
+   */
+  private static abstract class TestFuture<T> {
+    final Future<T> future;
+    final String operationType;
+    final OperationInfo opInfo;
+
+    TestFuture(Future<T> future, String operationType, OperationInfo opInfo) {
+      this.future = future;
+      this.operationType = operationType;
+      this.opInfo = opInfo;
+    }
+
+    /**
+     * Generate a name for the tested operation
+     * @return the operation name
+     */
+    String getOperationName() {
+      return operationType + "-" + opInfo.operationId;
+    }
+
+    /**
+     * Return the value inside the future or throw an {@link AssertionError} if an exception
+     * occurred.
+     * @return the value inside the future
+     */
+    T get() {
+      try {
+        return this.future.get();
+      } catch (Exception e) {
+        e.printStackTrace();
+        Assert.fail("Exception occured in operation: " + getOperationName() + ", exception: " + e);
+        return null;
+      }
+    }
+
+    /**
+     * Implement any testing logic here.
+     */
+    abstract void check();
+  }
+
+  /**
+   * A callback for router operations that starts the next operation in the chain after completion.
+   * The user can define a custom action on the result of the operation by overriding the {@code action()} method.
+   * @param <T> The callback's result type
+   */
+  private class TestCallback<T> implements Callback<T> {
+    final OperationInfo opInfo;
+    final boolean expectError;
+
+    TestCallback(OperationInfo opInfo, boolean expectError) {
+      this.opInfo = opInfo;
+      this.expectError = expectError;
+    }
+
+    @Override
+    public void onCompletion(T result, Exception exception) {
+      if (exception != null && !expectError) {
+        opInfo.latch.countDown();
```

Comment:
This block will complete the operation chain if an unexpected error occurs in an operation.